pkg/volume/csi/csi_mounter.go

```diff
@@ -0,0 +1,204 @@
+	return nil
+}
+
+func (c *csiMountMgr) SetUp(fsGroup *int64) error {
+	return c.SetUpAt(c.GetPath(), fsGroup)
+}
+
+func (c *csiMountMgr) SetUpAt(dir string, fsGroup *int64) error {
+	glog.V(4).Infof(log("Mounter.SetUpAt(%s)", dir))
+
+	ctx, cancel := grpctx.WithTimeout(grpctx.Background(), csiTimeout)
+	defer cancel()
+
+	csi := c.csiClient
+	pvName := c.spec.PersistentVolume.GetName()
+
+	// ensure version is supported
+	if err := csi.AssertSupportedVersion(ctx, csiVersion); err != nil {
+		glog.Errorf(log("failed to assert version: %v", err))
+		return err
+	}
+
+	// TODO (vladimirvivien) make assertion calls private inside client.
+	// assert volume attachment capability from driver,
+	// if not supported, assume vol is not previously attached and not ready for mounting
+	if err := csi.AssertVolumePublishCapability(ctx); err != nil {
+		glog.Error(log("failed to assert volume publish capability: %v", err))
+		return err
+	}
+
+	// search for attachment by VolumeAttachment.Spec.Source.PersistentVolumeName
+	if c.volumeInfo == nil {
+		attachList, err := c.k8s.StorageV1alpha1().VolumeAttachments().List(meta.ListOptions{})
+		if err != nil {
+			glog.Error(log("failed to get volume attachments: %v", err))
+			return err
+		}
+
+		var attachment *storage.VolumeAttachment
+		for _, attach := range attachList.Items {
+			if attach.Spec.Source.PersistentVolumeName != nil &&
+				*attach.Spec.Source.PersistentVolumeName == pvName {
+				attachment = &attach
+				break
+			}
+		}
+
+		if attachment == nil {
+			glog.Error(log("unable to find VolumeAttachment with PV.name = %s", pvName))
+			return errors.New("no existing VolumeAttachment found")
+		}
+		c.volumeInfo = attachment.Status.AttachmentMetadata
+	}
+
+	//TODO (vladimirvivien) implement better AccessModes mapping between k8s and CSI
+	accessMode := api.ReadWriteOnce
+	if c.spec.PersistentVolume.Spec.AccessModes != nil {
+		accessMode = c.spec.PersistentVolume.Spec.AccessModes[0]
+	}
+
+	err := csi.NodePublishVolume(
+		ctx,
+		c.volumeID,
+		c.readOnly,
+		dir,
+		accessMode,
+		c.volumeInfo,
+		"ext4", //TODO needs to be sourced from PV or somewhere else
+	)
+
+	if err != nil {
+		glog.Errorf(log("Mounter.Setup failed: %v", err))
+		return err
+	}
+	glog.V(4).Infof(log("successfully mounted %s", dir))
+
+	return nil
+}
+
+func (c *csiMountMgr) GetAttributes() volume.Attributes {
+	return volume.Attributes{
+		ReadOnly:        c.readOnly,
+		Managed:         !c.readOnly,
+		SupportsSELinux: false,
+	}
+}
+
+// volume.Unmounter methods
+var _ volume.Unmounter = &csiMountMgr{}
+
+func (c *csiMountMgr) TearDown() error {
+	return c.TearDownAt(c.GetPath())
+}
+func (c *csiMountMgr) TearDownAt(dir string) error {
+	glog.V(4).Infof(log("Unmounter.TearDown(%s)", dir))
+
+	// extract driverName and volID from path
+	base, volID := path.Split(dir)
+	volID = kstrings.UnescapeQualifiedNameForDisk(volID)
+	driverName := path.Base(base)
+
+	if !isDriverNameValid(driverName) {
```

Comment:
nit: This is a weird place to validate the driver name. I would just assume it is correct, if it is not then the operation will just handle it self.

Reply:
You are right, there's no need for it since it's extracted from the directory path it should have been ok.